### PR TITLE
 [PEP 573] bpo-38787: Module State Access from C Extension Methods

### PR DESCRIPTION
**Pull request by ghost**
This is a WIP pull request for PEP 573.

The purpose of this PEP is to provide a way to access per-module state from methods belonging to classes defined in extension modules.

https://bugs.python.org/issue38787

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -2912,6 +2922,11 @@ PyType_FromSpecWithBases(PyType_Spec *spec, PyObject *bases)
     Py_INCREF(res->ht_qualname);
     type->tp_name = spec->name;
 
+    if (module) {
+        Py_INCREF(module);
```
Comment (scoder) [inline Objects/typeobject.c]: `Py_XINCREF()`?

File: Objects/typeobject.c [hunk 2]
```diff
@@ -3066,6 +3081,67 @@ PyType_GetSlot(PyTypeObject *type, int slot)
...
+    PyObject *mro = type->tp_mro;
+
+    assert(PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE));
+
+    while(1) {
+        if (PyType_GetSlot(type, slot) == func) {
+            break;
+        }
+
+        type = (PyTypeObject *)PyTuple_GetItem(mro, pos);
```
Comment (scoder) [inline Objects/typeobject.c]: What is the guarantee that prevents the two tuple item accesses (here and below) from running out of bounds?
Could that be worth a comment, at least?

File: Include/object.h [hunk 3]
```diff
@@ -177,6 +177,11 @@ typedef struct _typeobject PyTypeObject;
 /* PyTypeObject is defined in cpython/object.h */
 #endif
 
+typedef struct {
+    Py_ssize_t dict;
+    Py_ssize_t weaklist;
+} PyType_offsets;
+
```
Comment (encukou) [inline Include/object.h]: This should be unnecessary.

File: Include/methodobject.h [hunk 4]
```diff
@@ -12,8 +12,9 @@ extern "C" {
    for the latter. */
 
 PyAPI_DATA(PyTypeObject) PyCFunction_Type;
+PyAPI_DATA(PyTypeObject) PyCMethod_Type;
 
-#define PyCFunction_Check(op) (Py_TYPE(op) == &PyCFunction_Type)
+#define PyCFunction_Check(op) ((Py_TYPE(op) == &PyCFunction_Type) || (PyType_IsSubtype(Py_TYPE(op), &PyCFunction_Type)))
```
Comment (encukou) [inline Include/methodobject.h:17]: Can we use the existing `PyCFunction_Type` for methods?
The `METH_METHOD` flag should signal that the underlying struct is `PyCMethodObject`; you'll also want to set `tp_itemsize` and the use `PyObject_GC_NewVar`.

File: Include/methodobject.h [hunk 5]
```diff
@@ -23,10 +24,12 @@ typedef PyObject *(*_PyCFunctionFastWithKeywords) (PyObject *,
                                                    PyObject *const *, Py_ssize_t,
                                                    PyObject *);
 typedef PyObject *(*PyNoArgsFunction)(PyObject *);
+typedef PyObject *(*PyCMethod)(PyObject *, PyTypeObject *, PyObject *, PyObject *);
 
 PyAPI_FUNC(PyCFunction) PyCFunction_GetFunction(PyObject *);
 PyAPI_FUNC(PyObject *) PyCFunction_GetSelf(PyObject *);
 PyAPI_FUNC(int) PyCFunction_GetFlags(PyObject *);
+PyAPI_FUNC(PyTypeObject *) PyCMethod_GetClass(PyObject *);
```
Comment (encukou) [inline Include/methodobject.h]: This looks unnecessary; the PEP only mentions `PyCFunction_GET_CLASS`.

File: Include/object.h [hunk 6]
```diff
@@ -197,6 +202,18 @@ PyAPI_FUNC(PyObject*) PyType_FromSpecWithBases(PyType_Spec*, PyObject*);
 #if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= 0x03040000
 PyAPI_FUNC(void*) PyType_GetSlot(struct _typeobject*, int);
 #endif
+#if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= 0x03080000
+PyAPI_FUNC(PyObject*) PyType_FromModuleAndSpec(PyObject *, PyType_Spec *, PyObject *);
```
Comment (encukou) [inline Include/object.h:201]: Oops, this is missing from the PEP's [summary](https://www.python.org/dev/peps/pep-0573/#summary-of-api-changes-and-additions).

File: Include/object.h [hunk 7]
```diff
@@ -197,6 +202,18 @@ PyAPI_FUNC(PyObject*) PyType_FromSpecWithBases(PyType_Spec*, PyObject*);
...
+#endif
+
+
+#ifndef Py_LIMITED_API
+/* The *real* layout of a type object when allocated on the heap */
+
+/* access macro to the members which are floating "behind" the object */
+#define PyHeapType_GET_MEMBERS(etype) \
+    ((PyMemberDef *)(((char *)etype) + Py_TYPE(etype)->tp_basicsize))
+#endif
```
Comment (encukou) [inline Include/object.h]: This is duplicates the definition in `Include/internal/object.h`.

File: Lib/test/test_importlib/extension/test_loader.py [hunk 8]
```diff
@@ -9,8 +9,10 @@
 import unittest
 import importlib.util
 import importlib
+import gc
 from test.support.script_helper import assert_python_failure
 
+
```
Comment (encukou) [inline Lib/test/test_importlib/extension/test_loader.py]: Unused import & unrelated whitespace change

File: Lib/test/test_io.py [hunk 9]
```diff
@@ -3496,8 +3496,9 @@ def __del__(self):
     def test_create_at_shutdown_without_encoding(self):
         rc, out, err = self._check_create_at_shutdown()
         if err:
-            # Can error out with a RuntimeError if the module state
+            # Can error out with a Lookup if the encoding
```
Comment (encukou) [inline Lib/test/test_io.py]: LookupError?

This might be worth a separate PR.

File: Modules/_testmultiphase.c [hunk 10]
```diff
@@ -4,6 +4,22 @@
 
 #include "Python.h"
 
+/* State for testing access from type slot methods */
+
```
Comment (encukou) [inline Modules/_testmultiphase.c:8]:
```suggestion
```

File: Modules/_testmultiphase.c [hunk 11]
```diff
@@ -4,6 +4,22 @@
...
+/*[clinic input]
+module _testmultiphase
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=a11493b5ab5f8846]*/
+
+/*[clinic input]
+class _testmultiphase.Example "ExampleObject *" "!Example"
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=c23a06dc20bc1f6f]*/
```
Comment (encukou) [inline Modules/_testmultiphase.c]: I believe you can combine these blocks.

File: Modules/_testmultiphase.c [hunk 12]
```diff
@@ -98,10 +126,95 @@ static PyType_Spec Example_Type_spec = {
...
+    }
+    Py_INCREF(retval);
+    return retval;
+}
+
+/*[clinic input]
+_testmultiphase.StateAccessType.get_instance_count
+
+
+This method returns module of the defining class.
```
Comment (encukou) [inline Modules/_testmultiphase.c]:
```suggestion
This method returns the number of existing instances.
```

File: Objects/descrobject.c [hunk 13]
```diff
@@ -124,7 +124,11 @@ classmethod_get(PyMethodDescrObject *descr, PyObject *obj, PyObject *type)
                      ((PyTypeObject *)type)->tp_name);
         return NULL;
     }
-    return PyCFunction_NewEx(descr->d_method, type, NULL);
+    if (descr->d_method->ml_flags & METH_METHOD) {
+        return PyCMethod_New(descr->d_method, type, NULL, descr->d_common.d_type);
+    } else {
+        return PyCFunction_NewEx(descr->d_method, type, NULL);
+    }
```
Comment (encukou) [inline Objects/descrobject.c]: This should be much simpler, since `PyCFunction_NewEx(...)` is defined as `PyCMethod_New(..., NULL)`.

File: Objects/methodobject.c [hunk 14]
```diff
@@ -60,11 +70,44 @@ PyCFunction_NewEx(PyMethodDef *ml, PyObject *self, PyObject *module)
...
+    PyCFunctionObject *op = NULL;
+
+    if (ml->ml_flags & METH_METHOD) {
+        PyCMethodObject *om;
+        if (ml->ml_flags & (METH_NOARGS | METH_O | METH_CLASS | METH_STATIC)) {
+            PyErr_SetString(PyExc_SystemError,
+                            "METH_METHOD cannot be used with METH_NOARGS, "
+                            "METH_O, METH_CLASS, nor METH_STATIC");
+            return NULL;
+        }
```
Comment (encukou) [inline Objects/methodobject.c]: This looks like it would allow `(METH_VARARGS | METH_METHOD)` or `METH_FASTCALL | METH_METHOD`

File: Objects/typeobject.c (hunk 2 above)
Comment (encukou) [reply Objects/typeobject.c]: A `for` loop might be better: first use `PyTuple_Size` to also check the type, then use the faster `PyTuple_GET_ITEM`.

File: Modules/_testmultiphase.c [hunk 15]
```diff
@@ -98,10 +126,95 @@ static PyType_Spec Example_Type_spec = {
...
     0,
-    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC | Py_TPFLAGS_BASETYPE,
     Example_Type_slots
 };
 
+
+/*[clinic input]
+_testmultiphase.StateAccessType.__init__
+    cls: defining_class
```
Comment (encukou) [inline Modules/_testmultiphase.c]: Could you also test with another slot, not only `__init__`?

File: Tools/clinic/clinic.py [hunk 16]
```diff
@@ -1021,6 +1053,19 @@ def parser_body(prototype, *fields, declarations=''):
             if parses_keywords:
                 assert parses_positional
 
+            if requires_defining_class:
+                fields.insert(0, normalize_snippet("""
+                    PyTypeObject *cls;
+
+                    cls = PyType_DefiningTypeFromSlotFunc(Py_TYPE(self),
```
Comment (encukou) [inline Tools/clinic/clinic.py]: Is it always named `cls`?

**Review by encukou — COMMENTED** [review]
Lots of comments on first glance:
